### PR TITLE
feat (apptoken): app token create repository

### PR DESCRIPTION
**Pull request by uruemu**

### CODE REVIEW

File: internal/apptoken/validation.go [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package apptoken
+
+import "context"
+
+// ValidateAppTokenGrants will ensure that the apptokens grants don't exceed the
+// grants of the user
+func ValidateAppTokenGrants(ctx context.Context, gf grantFinder, createdByUserId string, appTokenGrants []string) error {
+	return nil
```
Comment (cameronperera) [inline internal/apptoken/validation.go:8]: Is the plan to implement this in a later PR?

File: internal/apptoken/repository_create_test.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+					Description:    "test-description",
+					ScopeId:        testOrg.GetPublicId(),
+				},
+			},
+			wantGrants: []*apptoken.AppTokenGrant{
+				{
+					AppTokenGrant: &store.AppTokenGrant{},
+				},
+			},
+		},
```
Comment (cameronperera) [inline internal/apptoken/repository_create_test.go:77]: We need to add error handling tests here as well correct?

File: internal/apptoken/repository.go [hunk 3]
```diff
@@ -48,6 +56,31 @@ func NewRepository(ctx context.Context, r db.Reader, w db.Writer, kms *kms.Kms,
...
+func (userHistory) TableName() string {
+	return "iam_user_hst"
+}
+
+// ResolveUserHistoryId will lookup the userId and return its user history id
+func (r *Repository) ResolveUserHistoryId(ctx context.Context, userId string) (string, error) {
+	const op = "apptoken.(Repository).ResolveUserHistoryId"
+
+	switch {
+	case userId == "":
```
Comment (jimlambrt) [inline internal/apptoken/repository.go:78]: should we write a test for this?

File: internal/apptoken/repository.go [hunk 4]
```diff
@@ -48,6 +56,31 @@ func NewRepository(ctx context.Context, r db.Reader, w db.Writer, kms *kms.Kms,
...
+		return "", errors.New(ctx, errors.InvalidParameter, op, "missing user id")
+	}
+	var hst userHistory
+	if err := r.reader.LookupWhere(ctx, &hst, "public_id = ?", []any{userId}, db.WithOrder("DESC valid_range")); err != nil {
+		return "", errors.Wrap(ctx, err, op)
```
Comment (jimlambrt) [inline internal/apptoken/repository.go:83]: should we write a test for this error?

File: internal/apptoken/repository_create.go [hunk 5]
```diff
@@ -2,3 +2,124 @@
...
+		if err != nil {
+			return nil, nil, errors.Wrap(ctx, err, op)
+		}
+		grants = append(grants, &grant)
+	}
+
+	// factory supports options: WithName and WithDescription
+	appT, err := NewAppToken(ctx, scopeId, expTime, createdByUserId, opt...)
+	if err != nil {
+		return nil, nil, errors.Wrap(ctx, err, op)
```
Comment (jimlambrt) [inline internal/apptoken/repository_create.go:49]: this might be an easy err to test for?

File: internal/apptoken/repository_create.go [hunk 6]
```diff
@@ -2,3 +2,124 @@
...
+	appTId, err := newAppTokenId(ctx)
+	if err != nil {
+		return nil, nil, errors.Wrap(ctx, err, op)
+	}
+
+	appT.PublicId = appTId
+
+	opLogWrapper, err := r.kms.GetWrapper(ctx, scopeId, kms.KeyPurposeOplog)
+	if err != nil {
+		return nil, nil, errors.Wrap(ctx, err, op)
```
Comment (jimlambrt) [inline internal/apptoken/repository_create.go:60]: might be easy to write a test for this err.

**Review by jimlambrt — CHANGES_REQUESTED** [review]
A few questions/suggestions.  I also pushed a fixup that you may want to review.

File: internal/apptoken/repository_create_test.go (hunk 2 above)
Comment (uruemu) [reply internal/apptoken/repository_create_test.go:77]: Added error handling tests

File: internal/apptoken/repository_create.go [hunk 7]
```diff
@@ -2,3 +2,124 @@
...
+	"github.com/hashicorp/boundary/internal/oplog"
+	"github.com/hashicorp/boundary/internal/perms"
+)
+
+// CreateAppToken will create an apptoken in the repository and return the written apptoken
+// Takes in grant string.  Options supported: WithName, WithDescription
+func (r *Repository) CreateAppToken(ctx context.Context, scopeId string, expTime time.Time, createdByUserId string, grantsStr []string, opt ...Option) (*AppToken, []*AppTokenGrant, error) {
+	const op = "apptoken.(Repository).CreateAppToken"
+
+	switch {
```
[the remark this answers is not in the document]
Comment (uruemu) [reply internal/apptoken/repository_create.go:22]: Added tests

File: internal/apptoken/repository_create.go [hunk 8]
```diff
@@ -2,3 +2,124 @@
...
+	}
+
+	grants := make([]*perms.Grant, 0, len(grantsStr))
+	for _, grantStr := range grantsStr {
+		// Validate that the grant parses successfully. Note that we fake the scope
+		// here to avoid a lookup as the scope is only relevant at actual ACL
+		// checking time and we just care that it parses correctly.
+		const fakeScopeId = "o_abcd1234"
+		grant, err := perms.Parse(ctx, fakeScopeId, grantStr)
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (uruemu) [reply internal/apptoken/repository_create.go:40]: Yup, added tests

File: internal/apptoken/repository_create.go [hunk 9]
```diff
@@ -2,3 +2,124 @@
...
+
+	appTokenMetadata, err := appT.oplog(ctx, oplog.OpType_OP_TYPE_CREATE)
+	if err != nil {
+		return nil, nil, errors.Wrap(ctx, err, op)
+	}
+
+	var retAppToken *AppToken
+	var retAppTokenGrants []*AppTokenGrant
+	_, err = r.writer.DoTx(ctx, db.StdRetryCnt, db.ExpBackoff{}, func(r db.Reader, w db.Writer) error {
+		//TODO: Add ticket type for AppTOken in SQL
```
[the remark this answers is not in the document]
Comment (uruemu) [reply internal/apptoken/repository_create.go]: Done

File: internal/apptoken/repository.go (hunk 3 above)
Comment (uruemu) [reply internal/apptoken/repository.go:78]: Done

File: internal/apptoken/repository.go (hunk 4 above)
Comment (uruemu) [reply internal/apptoken/repository.go:83]: Done

File: internal/apptoken/repository_create.go (hunk 5 above)
Comment (uruemu) [reply internal/apptoken/repository_create.go:49]: The error paths here are not reachable due to the validations above so I won't be able to test the error cases but the sucess path is covered in the tests

File: internal/apptoken/repository_create.go (hunk 6 above)
Comment (uruemu) [reply internal/apptoken/repository_create.go:60]: Added tests

**Review by jimlambrt — APPROVED** [review]
ty!

**Review by cameronperera — APPROVED** [review]
LGTM

**Pull request merged**